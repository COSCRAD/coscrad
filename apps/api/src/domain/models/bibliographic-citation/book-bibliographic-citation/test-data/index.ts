export * from './build-book-bibliographic-citation-created';
