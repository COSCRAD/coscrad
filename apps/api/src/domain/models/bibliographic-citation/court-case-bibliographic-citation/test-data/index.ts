export * from './build-court-case-bibliographic-citation-created';
