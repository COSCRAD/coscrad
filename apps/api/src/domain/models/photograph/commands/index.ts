export * from './create-photograph';
export * from './photograph.commands.module';
