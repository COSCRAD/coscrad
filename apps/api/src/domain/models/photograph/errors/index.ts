export * from './invalid-mime-type-for-photograph.error';
