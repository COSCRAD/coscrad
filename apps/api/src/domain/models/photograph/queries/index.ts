export * from './photograph-query-repository.interface';
