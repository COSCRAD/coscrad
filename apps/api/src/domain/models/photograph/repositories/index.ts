export * from './arango-photograph-query-repository';
