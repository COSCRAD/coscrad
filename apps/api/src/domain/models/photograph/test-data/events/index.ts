export * from './get-photograph-test-event-builders';
