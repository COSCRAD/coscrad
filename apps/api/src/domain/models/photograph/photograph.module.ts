import { CommandModule } from '@coscrad/commands';
import { Module } from '@nestjs/common';
import { CreatePhotograph, PhotographCommandsModule, PhotographCreated } from '.';
import { CommandInfoService } from '../../../app/controllers/command/services/command-info-service';
import { PhotographController } from '../../../app/controllers/resources/photograph.controller';
import { ConsoleCoscradCliLogger } from '../../../coscrad-cli/logging';
import { IdGenerationModule } from '../../../lib/id-generation/id-generation.module';
import { ArangoConnectionProvider } from '../../../persistence/database/arango-connection.provider';
import { PersistenceModule } from '../../../persistence/persistence.module';
import { DynamicDataTypeModule } from '../../../validation';
import { EventModule } from '../../common';
import { CreatePhotographCommandHandler } from './commands/create-photograph/create-photograph.command-handler';
import { PhotographCreatedEventHandler } from './commands/create-photograph/photograph-created.event-handler';
import { Photograph } from './entities/photograph.entity';
import { PHOTOGRAPH_QUERY_REPOSITORY_TOKEN } from './queries';
import { PhotographQueryService } from './queries/photograph-query.service';
import { ArangoPhotographQueryRepository } from './repositories';

@Module({
    imports: [
        PersistenceModule,
        DynamicDataTypeModule,
        CommandModule,
        IdGenerationModule,
        EventModule,
        PhotographCommandsModule,
    ],
    controllers: [PhotographController],
    providers: [
        CommandInfoService,
        PhotographQueryService,
        {
            provide: PHOTOGRAPH_QUERY_REPOSITORY_TOKEN,
            useFactory: (arangoConnectionProvider: ArangoConnectionProvider) =>
                new ArangoPhotographQueryRepository(
                    arangoConnectionProvider,
                    new ConsoleCoscradCliLogger()
                ),
            inject: [ArangoConnectionProvider],
        },
        CreatePhotograph,
        CreatePhotographCommandHandler,
        // Data classes
        ...[
            // Domain Model
            Photograph,
            // Events
            PhotographCreated,
        ].map((Ctor) => ({
            provide: Ctor,
            useValue: Ctor,
        })),
        PhotographCreatedEventHandler,
    ],
    exports: [PhotographQueryService],
})
export class PhotographModule {}
