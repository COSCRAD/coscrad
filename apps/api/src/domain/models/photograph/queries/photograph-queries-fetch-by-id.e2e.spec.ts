import {
    AggregateType,
    CoscradUserRole,
    HttpStatusCode,
    IDetailQueryResult,
    IPhotographViewModel,
    LanguageCode,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes from '../../../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../../../app/controllers/__tests__/setUpIntegrationTest';
import getValidAggregateInstanceForTest from '../../../../domain/__tests__/utilities/getValidAggregateInstanceForTest';
import { clonePlainObjectWithOverrides } from '../../../../lib/utilities/clonePlainObjectWithOverrides';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from '../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestDataInFlatFormat from '../../../../test-data/buildTestDataInFlatFormat';
import { buildTestInstance } from '../../../../test-data/utilities';
import { buildMultilingualTextWithSingleItem } from '../../../common/build-multilingual-text-with-single-item';
import { AggregateId } from '../../../types/AggregateId';
import { assertQueryResult, assertResourceHasContributionFor } from '../../__tests__';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import {
    IPhotographQueryRepository,
    PHOTOGRAPH_QUERY_REPOSITORY_TOKEN,
} from '../../photograph/queries';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { ContributionSummary } from '../../user-management/contributor';
import { CoscradUserGroup } from '../../user-management/group/entities/coscrad-user-group.entity';
import { CoscradUserWithGroups } from '../../user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../user-management/user/entities/user/coscrad-user.entity';
import { FullName } from '../../user-management/user/entities/user/full-name.entity';
import { PhotographViewModel } from './photograph.view-model';

// Set up endpoints: index endpoint, id endpoint
const indexEndpoint = `/resources/photographs`;

const buildDetailEndpoint = (id: AggregateId) => `${indexEndpoint}/${id}`;

// We require an existing user for ACL tests
const dummyQueryUserId = buildDummyUuid(4);

const { user: users, userGroup: userGroups } = buildTestDataInFlatFormat();

const dummyUser = (users as CoscradUser[])[0].clone({
    authProviderUserId: `auth0|${dummyQueryUserId}`,
    id: dummyQueryUserId,
    roles: [CoscradUserRole.viewer],
});

const dummyGroup = (userGroups as CoscradUserGroup[])[0].clone({ userIds: [dummyUser.id] });

const _dummyUserWithGroups = new CoscradUserWithGroups(dummyUser, [dummyGroup]);

// Set up test data (use event sourcing to set up state)
const photographTitle = `Photograph Title`;

const languageCodeForTitle = LanguageCode.Haida;

const photographId = buildDummyUuid(1);

const photographCompositeId = {
    type: AggregateType.photograph,
    id: photographId,
};

const existingMediaItem = getValidAggregateInstanceForTest(AggregateType.mediaItem).clone({
    id: buildDummyUuid(55),
});

const testPhotographer = 'Jeff Thomas';

const photographHeightPx = 600;

const photographWidthPx = 800;

const dummyContributorFirstName = 'Dumb';

const dummyContributorLastName = 'McContributor';

const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor).clone({
    fullName: new FullName({
        firstName: dummyContributorFirstName,
        lastName: dummyContributorLastName,
    }),
});

const dummyPhotograph = buildTestInstance(PhotographViewModel, {
    ...photographCompositeId,
    name: buildMultilingualTextWithSingleItem(photographTitle, languageCodeForTitle),
    mediaItemId: existingMediaItem.id,
    photographer: testPhotographer,
    contributions: [
        buildTestInstance(ContributionSummary, {
            contributorIds: [dummyContributor.id],
        }),
    ],
    heightPx: photographHeightPx,
    widthPx: photographWidthPx,
});

const targetPhotographView = clonePlainObjectWithOverrides(dummyPhotograph, {
    isPublished: true,
    contributions: [
        buildTestInstance(ContributionSummary, {
            contributorIds: [dummyContributor.id],
        }),
    ],
});

describe(`when querying for a photograph: fetch by Id`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let photographQueryRepository: IPhotographQueryRepository;

    const setItUp = async (userWithGroups: CoscradUserWithGroups) => {
        ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
            {
                ARANGO_DB_NAME: testDatabaseName,
                BASE_URL: 'https://www.photograph-queries-by-id.com',
                GLOBAL_PREFIX: 'myapi',
            },
            {
                testUserWithGroups: userWithGroups,
            }
        ));

        photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
    };

    // let eventPublisher: ICoscradEventPublisher;
    beforeEach(async () => {
        await testRepositoryProvider.testSetup();

        await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
        });

        describe(`when there is a photograph with the given Id`, () => {
            describe(`when a photograph is published`, () => {
                beforeEach(async () => {
                    /**
                     * It is important that we do this before allowing the event
                     * handlers to run, as they will only add contributions
                     * if the corresponding contributor exists.
                     */
                    await testRepositoryProvider
                        .getContributorRepository()
                        .create(dummyContributor);

                    await databaseProvider.getDatabaseForCollection('photograph__VIEWS').clear();

                    await photographQueryRepository.createMany([targetPhotographView]);
                });

                /**
                 * TODO We need a contract test that will warn us if we potentially
                 * break the client via breaking changes to the API.
                 */
                it('should return the expected result', async () => {
                    const res = await request(app.getHttpServer()).get(
                        buildDetailEndpoint(photographId)
                    );

                    expect(res.status).toBe(httpStatusCodes.ok);

                    const result = res.body as IDetailQueryResult<PhotographViewModel>;

                    expect(result.id).toBe(photographId);

                    // We don't expose actions to non-admin users
                    expect(result.actions).toEqual([]);

                    assertResourceHasContributionFor(dummyContributor, result);
                });
            });

            describe(`when a photograph is not published`, () => {
                beforeEach(async () => {
                    // note that there is no publication event in this event history

                    // TODO: we need to check that contributors come through
                    await photographQueryRepository.createMany([
                        clonePlainObjectWithOverrides(targetPhotographView, {
                            isPublished: false,
                        }),
                    ]);
                });

                // We pretend the resource does not exist when the user
                // does not have access to this photograph
                it(`should return not found (404)`, async () => {
                    const res = await request(app.getHttpServer()).get(
                        buildDetailEndpoint(photographId)
                    );

                    expect(res.status).toBe(httpStatusCodes.notFound);
                });
            });
        });

        describe(`when there is no photograph with the given Id`, () => {
            it(`should return not found (404)`, async () => {
                const res = await request(app.getHttpServer()).get(
                    buildDetailEndpoint(buildDummyUuid(467))
                );

                expect(res.status).toBe(httpStatusCodes.notFound);
            });
        });
    });

    describe(`when the user is authenticated`, () => {
        describe(`when the user is a coscrad admin`, () => {
            beforeAll(async () => {
                await setItUp(
                    new CoscradUserWithGroups(
                        dummyUser.clone({
                            roles: [CoscradUserRole.superAdmin],
                        }),
                        []
                    )
                );

                photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
            });

            describe(`when there is a photograph with the given Id`, () => {
                describe(`when the photograph is published`, () => {
                    it(`should return the expected result`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await photographQueryRepository.createMany([targetPhotographView]);
                            },
                            endpoint: buildDetailEndpoint(photographId),
                            expectedStatus: httpStatusCodes.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IPhotographViewModel>
                            ) => {
                                expect(body.actions).not.toEqual([]);
                            },
                        });
                    });
                });

                describe(`when the photograph is not published`, () => {
                    it(`should return the expected result`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await photographQueryRepository.createMany([
                                    clonePlainObjectWithOverrides(targetPhotographView, {
                                        isPublished: false,
                                    }),
                                ]);
                            },
                            endpoint: buildDetailEndpoint(photographId),
                            // admin can see unpublished resources
                            expectedStatus: HttpStatusCode.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IPhotographViewModel>
                            ) => {
                                expect(body.actions).not.toEqual([]);
                            },
                        });
                    });
                });
            });

            describe(`when there is no photograph with the given Id`, () => {
                it(`should return not found (404)`, async () => {
                    await assertQueryResult({
                        app,
                        seedInitialState: async () => {
                            Promise.resolve();
                        },
                        endpoint: buildDetailEndpoint('missing123'),
                        expectedStatus: HttpStatusCode.notFound,
                    });
                });
            });
        });

        describe(`when the user is a project admin`, () => {
            beforeAll(async () => {
                await setItUp(
                    new CoscradUserWithGroups(
                        dummyUser.clone({
                            roles: [CoscradUserRole.projectAdmin],
                        }),
                        []
                    )
                );

                photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
            });

            describe(`when there is a photograph with the given Id`, () => {
                describe(`when the photograph is published`, () => {
                    it(`should return the expected result`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await photographQueryRepository.createMany([targetPhotographView]);
                            },
                            endpoint: buildDetailEndpoint(photographId),
                            expectedStatus: httpStatusCodes.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IPhotographViewModel>
                            ) => {
                                expect(body.actions).not.toEqual([]);

                                /**
                                 * Note that we do this once and only once as
                                 * a contract test. We choose an admin case
                                 * because we want to capture the format of
                                 * the actions property as well.
                                 */
                                expect(body).toMatchSnapshot();
                            },
                        });
                    });
                });

                describe(`when the photograph is not published`, () => {
                    it(`should return the expected result`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await photographQueryRepository.createMany([
                                    clonePlainObjectWithOverrides(targetPhotographView, {
                                        isPublished: false,
                                    }),
                                ]);
                            },
                            endpoint: buildDetailEndpoint(photographId),
                            // admin can see unpublished resources
                            expectedStatus: HttpStatusCode.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IPhotographViewModel>
                            ) => {
                                expect(body.actions).not.toEqual([]);
                            },
                        });
                    });
                });
            });

            describe(`when there is no photograph with the given Id`, () => {
                it(`should return not found (404)`, async () => {
                    await assertQueryResult({
                        app,
                        seedInitialState: async () => {
                            Promise.resolve();
                        },
                        endpoint: buildDetailEndpoint('missing123'),
                        expectedStatus: HttpStatusCode.notFound,
                    });
                });
            });
        });

        describe(`when the user is an ordinary viewer (non-admin)`, () => {
            beforeAll(async () => {
                await setItUp(
                    new CoscradUserWithGroups(
                        dummyUser.clone({
                            roles: [CoscradUserRole.viewer],
                        }),
                        []
                    )
                );

                photographQueryRepository = app.get(PHOTOGRAPH_QUERY_REPOSITORY_TOKEN);
            });

            describe(`when there is a photograph with the given ID`, () => {
                describe(`when the photograph is published`, () => {
                    it(`should return the expected result`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await photographQueryRepository.createMany([targetPhotographView]);
                            },
                            endpoint: buildDetailEndpoint(photographId),
                            expectedStatus: httpStatusCodes.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IPhotographViewModel>
                            ) => {
                                // actions should not be visible to a non-admin user
                                expect(body.actions).toEqual([]);
                            },
                        });
                    });
                });

                describe(`when the photograph is not published`, () => {
                    describe(`when the user is not in the query ACL`, () => {
                        it(`should return the expected result`, async () => {
                            await assertQueryResult({
                                app,
                                seedInitialState: async () => {
                                    await photographQueryRepository.createMany([
                                        clonePlainObjectWithOverrides(targetPhotographView, {
                                            isPublished: false,
                                            // empty
                                            accessControlList: new AccessControlList(),
                                        }),
                                    ]);
                                },
                                endpoint: buildDetailEndpoint(photographId),
                                expectedStatus: HttpStatusCode.notFound,
                            });
                        });
                    });

                    describe(`when the user has access via the query ACL`, () => {
                        describe(`when the user is not in the query ACL`, () => {
                            it(`should return the expected result`, async () => {
                                await assertQueryResult({
                                    app,
                                    seedInitialState: async () => {
                                        await photographQueryRepository.createMany([
                                            clonePlainObjectWithOverrides(targetPhotographView, {
                                                isPublished: false,
                                                // empty
                                                accessControlList:
                                                    new AccessControlList().allowUser(dummyUser.id),
                                            }),
                                        ]);
                                    },
                                    endpoint: buildDetailEndpoint(photographId),
                                    expectedStatus: HttpStatusCode.ok,
                                    checkResponseBody: async (
                                        body: IDetailQueryResult<IPhotographViewModel>
                                    ) => {
                                        expect(body.actions).toEqual([]);
                                    },
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});
