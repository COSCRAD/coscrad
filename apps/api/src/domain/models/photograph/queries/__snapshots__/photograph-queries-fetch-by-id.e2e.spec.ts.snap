// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a photograph: fetch by Id when the user is authenticated when the user is a project admin when there is a photograph with the given Id when the photograph is published should return the expected result 1`] = `
{
  "accessControlList": {
    "allowedGroupIds": [],
    "allowedUserIds": [],
  },
  "actions": [
    {
      "description": "Allow a user to view (but not edit) a given resource",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a UUID",
                "name": "UUID",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "the ID of the user who will be given permission to view this resource",
            "label": "userId",
            "name": "userId",
            "options": {
              "aggregateType": "user",
            },
            "type": "DYNAMIC_SELECT",
          },
        ],
        "prepopulatedFields": {
          "aggregateCompositeIdentifier": {
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
            "type": "photograph",
          },
        },
      },
      "label": "Grant Read Access to User",
      "type": "GRANT_RESOURCE_READ_ACCESS_TO_USER",
    },
    {
      "description": "Make a resource unpublished to the public",
      "form": {
        "fields": [],
        "prepopulatedFields": {
          "aggregateCompositeIdentifier": {
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
            "type": "photograph",
          },
        },
      },
      "label": "Unpublish Resource",
      "type": "UNPUBLISH_RESOURCE",
    },
    {
      "description": "creates a note about this particular resource",
      "form": {
        "fields": [
          {
            "constraints": [],
            "description": "system-wide unique identifier for the resource about which we are making a note",
            "label": "CompositeIdentifier",
            "name": "resourceCompositeIdentifier",
            "type": "JSON_INPUT",
          },
          {},
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the note",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language in which you are writing the note",
            "label": "language code",
            "name": "languageCode",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {
          "aggregateCompositeIdentifier": {
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
            "type": "photograph",
          },
        },
      },
      "label": "Create Note",
      "type": "CREATE_NOTE_ABOUT_RESOURCE",
    },
    {
      "description": "connects two resources with a note and context",
      "form": {
        "fields": [
          {
            "constraints": [],
            "description": "system-wide unique identifier for the resource to which you are making a connection",
            "label": "to member composite identifier",
            "name": "toMemberCompositeIdentifier",
            "type": "JSON_INPUT",
          },
          {},
          {
            "constraints": [],
            "description": "system-wide unique identifier for the resource from which you are making a connection",
            "label": "from member composite identifier",
            "name": "fromMemberCompositeIdentifier",
            "type": "JSON_INPUT",
          },
          {},
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the note",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language in which you are writing the note",
            "label": "language code",
            "name": "languageCode",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {
          "aggregateCompositeIdentifier": {
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
            "type": "photograph",
          },
        },
      },
      "label": "Connect Resources with Note",
      "type": "CONNECT_RESOURCES_WITH_NOTE",
    },
  ],
  "contributions": [
    {
      "fullName": "Dumb McContributor",
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110901",
    },
  ],
  "heightPx": 600,
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
  "imageUrl": "https://www.photograph-queries-by-id.com/myapi/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100055",
  "isPublished": true,
  "name": {
    "items": [
      {
        "languageCode": "hai",
        "role": "original",
        "text": "Photograph Title",
      },
    ],
  },
  "photographer": "Jeff Thomas",
  "tags": [],
  "widthPx": 800,
}
`;
